File: src/api/routes/events.py
from fastapi import APIRouter, Depends, Header, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from ...config import settings
from ...db.models import Event, Session
from ...db.session import get_db_session
from ...security.ingest_token import verify_ingest_token
from ..schemas import EventsBatchRequest

router = APIRouter(prefix="/events", tags=["events"])


@router.post("/batch")
async def ingest_events_batch(
    body: EventsBatchRequest,
    authorization: str | None = Header(default=None, alias="Authorization"),
    db: AsyncSession = Depends(get_db_session),
) -> dict:
    if not authorization or not authorization.startswith("Bearer "):
        raise HTTPException(status_code=401, detail="missing bearer token")
    token = authorization.split(" ", 1)[1]
    payload = verify_ingest_token(settings.ingest_secret, token)
    if payload.get("sid") != body.session_id:
        raise HTTPException(status_code=403, detail="session mismatch")

    # validate session exists
    res = await db.execute(select(Session.id).where(Session.id == body.session_id))
    if res.scalar_one_or_none() is None:
        raise HTTPException(status_code=404, detail="session not found")

    # insert events
    db.add_all(
        [
            Event(session_id=body.session_id, t_ms=e.t_ms, type=e.type, payload=e.payload)
            for e in body.events
        ]
    )
    await db.commit()
    return {"accepted": True, "count": len(body.events)}

File: worker/routes/events.ts
import { Hono } from 'hono';
import { zValidator } from '@hono/zod-validator';
import { drizzle } from 'drizzle-orm/d1';
import { eq } from 'drizzle-orm';
import { events, sessions } from '../db/schema';
import { eventsBatchRequestSchema } from '../lib/schemas';
import { verifyIngestToken, extractBearerToken } from '../lib/auth';

type Env = {
  DB: D1Database;
  INGEST_SECRET: string;
};

const eventsRouter = new Hono<{ Bindings: Env }>();

eventsRouter.post('/batch', zValidator('json', eventsBatchRequestSchema), async (c) => {
  const body = c.req.valid('json');
  const db = drizzle(c.env.DB);
  
  // 토큰 검증
  const authorization = c.req.header('Authorization') || null;
  const token = extractBearerToken(authorization);
  
  if (!token) {
    return c.json({ error: 'Missing bearer token' }, 401);
  }
  
  let payload;
  try {
    payload = await verifyIngestToken(c.env.INGEST_SECRET, token);
  } catch (error) {
    return c.json({ error: 'Invalid or expired token' }, 403);
  }
  
  if (payload.sid !== body.session_id) {
    return c.json({ error: 'Session mismatch' }, 403);
  }
  
  // 세션 확인
  const session = await db
    .select({ id: sessions.id })
    .from(sessions)
    .where(eq(sessions.id, body.session_id))
    .get();
  
  if (!session) {
    return c.json({ error: 'Session not found' }, 404);
  }
  
  // 이벤트 삽입
  if (body.events.length > 0) {
    await db.insert(events).values(
      body.events.map((e) => ({
        sessionId: body.session_id,
        tMs: e.t_ms,
        type: e.type,
        payload: e.payload as any,
      }))
    );
  }
  
  return c.json({ accepted: true, count: body.events.length });
});

export default eventsRouter;

